### PR TITLE
[SYCL] Add faster reduction implementations using atomic or/and intel…

### PR DESCRIPTION
**Pull request by v-klochkov**
…::reduce()

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

**Comment by romanovvlad** [comment]
@Pennycook could you review the patch?

**Comment by v-klochkov** [comment]
> I'm not sure about the names `fast_atomic` and `fast_reduce` going forward -- what we're really testing here is whether SYCL provides `native` atomics or reductions for those types (since a device is not required to guarantee that their implementation is fast).
> 
> I don't feel strongly enough about this to block merging this PR, but we might want to revisit this naming convention when tuning for additional platforms.

If for some device those 'native' atomics happen to work slowly, then the good move is to exclude such 'native' atomic from 'fast' atomics list and use different algorithm not using them. Such exclusion would require some additional changes and maybe dynamic/runtime checks on HOST (I did not think much about it yet).
Taking this idea into account using the word 'fast' seems reasonable, right?

**Comment by Pennycook** [comment]
> If for some device those 'native' atomics happen to work slowly, then the good move is to exclude such 'native' atomic from 'fast' atomics list and use different algorithm not using them. Such exclusion would require some additional changes and maybe dynamic/runtime checks on HOST (I did not think much about it yet).
> Taking this idea into account using the word 'fast' seems reasonable, right?

That's a good point.  I agree -- as long as we continue to update the logic and use these specializations only if the features are expected to improve performance, "fast" is a good name.

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -761,6 +785,48 @@ class __SYCL_EXPORT handler {
 #endif
   }
 
+  /// Implements parallel_for() accepting nd_range and 1 reduction variable
+  /// having 'read_write' access mode.
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:789]: Vlad, doesn't this comment already tell that (see the line 3 and 4 of this comment section)?

  /// Implements parallel_for() accepting nd_range and 1 reduction variable
  /// having 'read_write' access mode.
  /// This version uses fast sycl::atomic operations to update user's reduction
  /// variable at the end of each work-group work.

**Pull request merged**